alpenhorn/io/updownlock.py

```diff
@@ -0,0 +1,267 @@
+in one of the locked states, any other threads wanting to
+acquire the lock in that same state may do so (i.e. multiple
+threads can hold the lock in a particular state), but threads
+wanting the other lock state must wait for the lock to become
+unlocked before they can aquire the lock.
+
+Alpenhorn uses this lock to prevent race conditions while
+modifying the directory tree of a StorageNode: threads which
+are in the process of creating directories must do so while
+holding the up lock and threads which are in the process of
+removing directories must do so while holding the down lock.
+
+The lock ensures only one operation (creation or removal) can
+happen at any given time.
+"""
+
+import time
+import threading
+from collections import defaultdict
+
+
+class _UpDownLock:
+    """UpDownLock internals.
+
+    Used to prevent a reference loop.
+    """
+
+    __slots__ = ["count", "_lock", "_owners", "_is_unlocked"]
+
+    def __init__(self) -> None:
+        # This tracks the state of the lock:
+        #  > 0: the lock is in the "up" state
+        #  = 0: the lock is unlocked
+        #  < 0: the lock is in the "down" state
+        #
+        # The absolute value indicates how many times
+        # the lock is held
+        self.count = 0
+
+        # Protects self._count
+        self._lock = threading.Lock()
+
+        # Dict of threads holding the lock.  Keys are the
+        # thread IDs.  Values are how many times that thread
+        # owns the lock.
+        self._owners = defaultdict(int)
+
+        # Is the lock unlocked?  Handles threads waiting for
+        # unlock.
+        self._is_unlocked = threading.Condition(threading.RLock())
+
+    def acquire(self, blocking: bool, timeout: float, is_down: bool) -> bool:
+        """Acquire the lock in the specified state."""
+
+        now = time.monotonic()
+        me = threading.get_ident()
+
+        # Try to acquire the lock
+        with self._lock:
+            if is_down:
+                ok_to_lock = self.count <= 0
+            else:
+                ok_to_lock = self.count >= 0
+
+            if ok_to_lock:
+                if is_down:
+                    self.count -= 1
+                else:
+                    self.count += 1
+                self._owners[me] += 1
+                return True
+
+            # If we already hold the lock in the other state, fail
+            if self._owners[me] > 0:
+                raise RuntimeError("Can't acquire both locking states.")
+
+        # If not blocking, fail
+        if not blocking:
+            return False
+
+        # Otherwise, wait until unlocked
+        with self._is_unlocked:
+            if timeout < 0:
+                self._is_unlocked.wait()
+            else:
+                end_at = None
+                notified = False
+
+                # Wait until either we're notified or else we timeout
+                while not notified and end_at is None or time.monotonic() < end_at:
```

Comment:
```suggestion
                while not notified and (end_at is None or time.monotonic() < end_at):
```
In Python, `and` binds more tightly than `or`.